qt/python/mantidqt/project/project.py

```diff
@@ -0,0 +1,134 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantidqt package
+#
+from __future__ import (absolute_import, division, print_function, unicode_literals)
+
+import os
+from qtpy.QtWidgets import QFileDialog, QMessageBox
+
+from mantid.api import AnalysisDataService, AnalysisDataServiceObserver
+from mantidqt.io import open_a_file_dialog
+from mantidqt.project.projectloader import ProjectLoader
+from mantidqt.project.projectsaver import ProjectSaver
+
+
+class Project(AnalysisDataServiceObserver):
+    def __init__(self):
+        super(Project, self).__init__()
+        # Has the project been saved, to Access this call .saved
+        self.__saved = True
+
+        # Last save locations
+        self.last_project_location = None
+
+        self.observeAll(True)
+
+        self.project_file_ext = ".mtdproj"
+
+    def __get_saved(self):
+        return self.__saved
+
+    saved = property(__get_saved)
+
+    def save(self):
+        """
+        The function that is called if the save button is clicked on the mainwindow
+        :return: None; if the user cancels
+        """
+        if self.last_project_location is None:
+            return self.save_as()
+        else:
+            # Actually save
+            workspaces_to_save = AnalysisDataService.getObjectNames()
+            project_saver = ProjectSaver(self.project_file_ext)
+            project_saver.save_project(directory=self.last_project_location, workspace_to_save=workspaces_to_save,
+                                       interfaces_to_save=None)
+            self.__saved = True
+
+    def save_as(self):
+        """
+        The function that is called if the save as... button is clicked on the mainwindow
+        :return: None; if the user cancels.
+        """
+        directory = self._get_directory_finder(accept_mode=QFileDialog.AcceptSave)
+
+        # If none then the user cancelled
+        if directory is None:
+            return
+
+        # todo: get a list of workspaces but to be implemented on GUI implementation
+        self.last_project_location = directory
+        workspaces_to_save = AnalysisDataService.getObjectNames()
+        project_saver = ProjectSaver(self.project_file_ext)
+        project_saver.save_project(directory=directory, workspace_to_save=workspaces_to_save, interfaces_to_save=None)
+        self.__saved = True
+
+    @staticmethod
+    def _get_directory_finder(accept_mode):
+        directory = None
+        # Check if it exists
+        first_pass = True
+        while first_pass or (not os.path.exists(directory) and os.path.exists(directory + (os.path.basename(directory)
```

Comment:
I was a little confused by this loop at first but now I see we are asking the user to point at a directory rather than a file and this checks if there is a project file in the directory.

I think for loading we should just ask the user to point at the `.mtdproj` file - I think this is generally what people would expect from a loading type dialog.

Reply:
The intention is to simulate a do while loop concept from other languages as python doesn't seem to have an elegant way of performing this I used a boolean, for the first pass of the while loop which flipped after the start.

Reply:
Changes made :+1: